### PR TITLE
Add product_deleted webhook

### PR DESCRIPTION
**Pull request by d-wysocki**
I want to merge this change because...I want to add `PRODUCT_DELETED` webhook

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/product/bulk_mutations/products.py [hunk 1]
```diff
@@ -102,6 +103,13 @@ def perform_mutation(cls, _root, info, ids, **data):
 
         return response
 
+    @classmethod
+    def bulk_action(cls, queryset):
+        manager = get_plugins_manager()
```
Comment (fowczarek) [inline saleor/graphql/product/bulk_mutations/products.py:108]: We should get the manager from context.

File: saleor/plugins/webhook/tests/test_webhook.py [hunk 2]
```diff
@@ -147,6 +147,18 @@ def test_product_updated(mocked_webhook_trigger, settings, product):
     )
 
 
+@mock.patch("saleor.plugins.webhook.plugin.trigger_webhooks_for_event.delay")
+def test_product_deleted(mocked_webhook_trigger, settings, product):
+    settings.PLUGINS = ["saleor.plugins.webhook.plugin.WebhookPlugin"]
+    manager = get_plugins_manager()
+    manager.product_deleted(product)
+
+    expected_data = generate_product_payload(product)
```
Comment (korycins) [inline saleor/plugins/webhook/tests/test_webhook.py:156]: Can we delete the product first? Then assign ID once again and generate the payload?
We should simulate that product was deleted for real

File: saleor/graphql/product/bulk_mutations/products.py (hunk 1 above)
Comment (korycins) [reply saleor/graphql/product/bulk_mutations/products.py:108]: I see mismatches here. For the bulk method you call `product_deleted` for products that will be deleted. But for default product delete mutation you call `product_deleted` after the product was deleted.

File: saleor/graphql/product/bulk_mutations/products.py [hunk 3]
```diff
@@ -94,14 +94,21 @@ def perform_mutation(cls, _root, info, ids, **data):
                 variant__in=variants, order__status=OrderStatus.DRAFT
             ).values_list("pk", flat=True)
         )
-
+        data["manager"] = info.context.plugins
```
Comment (korycins) [inline saleor/graphql/product/bulk_mutations/products.py:97]: Why not:
`response = super().perform_mutation(_root, info, ids, manager=info.context.plugins, **data)`

File: saleor/graphql/product/bulk_mutations/products.py [hunk 4]
```diff
@@ -94,14 +94,21 @@ def perform_mutation(cls, _root, info, ids, **data):
...
         order_models.OrderLine.objects.filter(pk__in=order_line_pks).delete()
 
         return response
 
+    @classmethod
+    def bulk_action(cls, queryset, manager):
+        products = [product for product in queryset]
+        queryset.delete()
+        for product in products:
+            manager.product_deleted(product)
```
Comment (korycins) [inline saleor/graphql/product/bulk_mutations/products.py:110]: AFAIK The product here losts relations to own variants. I would say that providing at least an ID or slug of variants is really important to proper sync between systems.
```
products = [product for product in Product.objects.filter(name__icontains="Juice")]
In [60]: products[0].variants.all()
Out[60]: <ProductVariantQueryset [<ProductVariant: 1l>, <ProductVariant: 2l>, <ProductVariant: 500ml>]>
Number of queries 1
In [61]: Product.objects.filter(name__icontains="Juice").delete()
In [62]: products[0].variants.all()
Out[62]: <ProductVariantQueryset []>
Number of queries 1
```

File: saleor/graphql/product/bulk_mutations/products.py [hunk 5]
```diff
@@ -87,21 +87,39 @@ class Meta:
...
+    def bulk_action(cls, queryset, manager, product_to_variant):
+        product_variant_map = defaultdict(list)
+        for product, variant in product_to_variant:
+            product_variant_map[product].append(variant)
+
+        products = [product for product in queryset]
+        queryset.delete()
+        for product in products:
+            variants = product_variant_map.get(product.id)
+            manager.product_deleted(product, variants)
```
Comment (maarcingebala) [inline saleor/graphql/product/bulk_mutations/products.py:121]: Here we'll send multiple webhooks in a loop. I think it's fine for now, but in the long run, wouldn't it be better to be able to return a list of items in a single webhook? I'm not sure how it works in other systems. @patrys Do you have any thoughts on this?

File: saleor/graphql/product/bulk_mutations/products.py [hunk 6]
```diff
@@ -87,21 +87,39 @@ class Meta:
...
         # get draft order lines for products
         order_line_pks = list(
             order_models.OrderLine.objects.filter(
-                variant__in=variants, order__status=OrderStatus.DRAFT
+                variant_id__in=variants_ids, order__status=OrderStatus.DRAFT
             ).values_list("pk", flat=True)
         )
-
-        response = super().perform_mutation(_root, info, ids, **data)
+        data["product_to_variant"] = product_to_variant
```
Comment (korycins) [inline saleor/graphql/product/bulk_mutations/products.py:101]: The same thing as previous :D 
Why not:
```
super().perform_mutation(...., manager=info.context.plugins, product_to_variant= product_to_variant, **data)
```

File: saleor/graphql/product/bulk_mutations/products.py (hunk 5 above)
Comment (korycins) [reply saleor/graphql/product/bulk_mutations/products.py:121]: @d-wysocki  Can you confirm that you fully tested it? I mean - did you subscribe to the webhook product_deleted and checked the payload which we send?

**Review by korycins — COMMENTED** [review]
@d-wysocki  Kindly reminder about saleor-docs ;)

File: saleor/graphql/product/tests/test_bulk_delete.py [hunk 7]
```diff
@@ -249,6 +249,34 @@ def test_delete_products(
...
+            for product in product_list
+        ]
+    }
+    response = staff_api_client.post_graphql(
+        query, variables, permissions=[permission_manage_products]
+    )
+    content = get_graphql_content(response)
+
+    assert content["data"]["productBulkDelete"]["count"] == 3
+    assert mocked_webhook_trigger.called
```
Comment (fowczarek) [inline saleor/graphql/product/tests/test_bulk_delete.py:277]: Maybe `assert_called_once_with` ?

File: saleor/graphql/product/bulk_mutations/products.py [hunk 8]
```diff
@@ -87,21 +87,39 @@ class Meta:
...
         return response
 
+    @classmethod
+    def bulk_action(cls, queryset, manager, product_to_variant):
+        product_variant_map = defaultdict(list)
+        for product, variant in product_to_variant:
+            product_variant_map[product].append(variant)
+
+        products = [product for product in queryset]
+        queryset.delete()
```
Comment (fowczarek) [inline saleor/graphql/product/bulk_mutations/products.py:124]: In this webhook, we send data like "publication_date", "is_published" and "currency". Those data are channel related. Should we return a list of "publication_date", "is_published" and "currency" per channel?

File: saleor/graphql/product/bulk_mutations/products.py (hunk 5 above)
Comment (d-wysocki) [reply saleor/graphql/product/bulk_mutations/products.py:127]: @korycins Yes, I tested it with mock server.

Comment (korycins) [reply saleor/graphql/product/bulk_mutations/products.py:127]: Could you post here (or in the PR description) the webhook payload which you got?

File: saleor/plugins/webhook/tests/test_webhook.py [hunk 9]
```diff
@@ -148,6 +149,25 @@ def test_product_updated(mocked_webhook_trigger, settings, product):
...
+
+    product = product
+    variants_id = product.variants.all().values_list("id", flat=True)
+    product_id = product.id
+    product.delete()
+    product.id = product_id
+
+    manager.product_deleted(product, product.variants.all())
+
+    expected_data = generate_product_deleted_payload(product, variants_id)
```
Comment (fowczarek) [inline saleor/plugins/webhook/tests/test_webhook.py:165]: I put ipdb here and check how `expected_data` look. `Variants` in `expected_data` are empty :(
image

Comment (fowczarek) [reply saleor/plugins/webhook/tests/test_webhook.py:165]: In `expected_data` are missing fields compare to `PRODUCT_FIELDS`

File: saleor/graphql/product/bulk_mutations/products.py (hunk 5 above)
Comment (d-wysocki) [reply saleor/graphql/product/bulk_mutations/products.py:127]:
```
b'[{"type": "Product", "id": "UHJvZHVjdDo3Mw==", "variants": [206, 207, 205], "private_metadata": {}, "metadata": {"vatlayer.code": "standard", "vatlayer.description": "standard"}, "name": "Carrot Juice", "description": {"blocks": [{"data": {"text": "Improve your eyesight the natural way with 100% pure, squeezed carrot juice. The sweet, orange nectar of Mother Earth."}, "type": "paragraph"}]}, "updated_at": "2021-01-26T12:06:38.764Z", "charge_taxes": true, "weight": "1.0:kg"}]'
```

File: saleor/graphql/product/bulk_mutations/products.py (hunk 8 above)
Comment (d-wysocki) [reply saleor/graphql/product/bulk_mutations/products.py:124]: In `product_created` and `product_updated` we didnt send this fields per channel.

File: saleor/plugins/webhook/tests/test_webhook.py (hunk 9 above)
Comment (d-wysocki) [reply saleor/plugins/webhook/tests/test_webhook.py:165]: Its should be good now. You can see payload from my mock server https://github.com/mirumee/saleor/pull/6794#discussion_r566707941

File: saleor/graphql/product/tests/test_bulk_delete.py (hunk 7 above)
Comment (d-wysocki) [reply saleor/graphql/product/tests/test_bulk_delete.py:277]: This webhook is called three times in this mutation. 
I already tested with params are called when this webhook is triggered in `/saleor/plugins/webhook/tests/test_webhook.py` `test_product_deleted`.

File: saleor/plugins/webhook/tests/test_webhook.py (hunk 9 above)
Comment (fowczarek) [reply saleor/plugins/webhook/tests/test_webhook.py:165]: Still same result (`variants` is empty).
I think we should add 2 asserts. 
1. product id is not empty
2. variants is not empty

**Review by korycins — APPROVED** [review]

File: saleor/plugins/webhook/tests/test_webhook.py (hunk 9 above)
Comment (patrys) [reply saleor/plugins/webhook/tests/test_webhook.py:166]: 3. `variants` is a list of GraphQL `ID`s and not a list of integers.

**Review by fowczarek — APPROVED** [review]

**Pull request merged**